orttraining/orttraining/python/training/ortmodule/__init__.py

```diff
@@ -11,27 +11,49 @@
 ################################################################################
 ONNX_OPSET_VERSION = 12
 MINIMUM_TORCH_VERSION_STR = '1.8.1'
-TORCH_CPP_BUILD_DIR = os.path.join(os.path.dirname(__file__),'torch_inline_extensions')
+
+# Use one of the available directories as Torch CPP extension in the following order:
+#    1) Path at listed at TORCH_EXTENSIONS_DIR environment variable
+#    2) Default Python package dir
+#    3) <Home directory>/.cache
+home_dir = os.path.expanduser("~")
+python_package_dir = os.path.dirname(__file__)
+torch_extensions_dir = os.environ.get('TORCH_EXTENSIONS_DIR')
+
+TORCH_CPP_BUILD_DIR = os.path.join(python_package_dir,'torch_inline_extensions')
+TORCH_CPP_BUILD_DIR_BACKUP = os.path.join(home_dir, '.cache', 'torch_ort_extensions')
+
+if torch_extensions_dir is not None and os.access(torch_extensions_dir, os.X_OK | os.W_OK):
+    TORCH_CPP_BUILD_DIR = torch_extensions_dir
+elif not os.access(python_package_dir, os.X_OK | os.W_OK):
```

Comment:
Usually os.access is the wrong function to use. It is designed for SUID programs. If you want to test if a path exists, you should use the https://linux.die.net/man/2/stat function.

Reply:
`python_package_dir` is assumed to exist. We need to check whether the current user can write to it

Smetimes we have docker image with `python_package_dir` with root-only access but training screipts are executed by other users. `os.access` will detect this scenario and another folder will be used instead